### PR TITLE
Refactors VV

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request

The thing other than ruining maps that I was working on
Refactors VV to use a more standard way of doing topic dropdown options rather than a huge if/else chain
Marking datums is now a right click option
Moves a few files around too/few procs

## Why It's Good For The Game

Makes it easier to add more VV dropdown options in the future, and moving href list keys to defines make misspelling them harder.

## Changelog
:cl:
add: Oh yeah also added a "return value of proccall" option for VV var editing.
refactor: View Variables has been refactored. It should now be easier to make VV dropdown options. 
/:cl:

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
that title scares me

**Comment by 81Denton** [comment]
At least it's not an untested kevinz webeditor PR 🤔

**Comment by 81Denton** [comment]
Shitposting aside the if/else chain removal looks really good

**Comment by Tlaltecuhtli** [comment]
will it have tgui?

**Comment by kevinz000** [comment]
@Tlaltecuhtli no TGUI, too laggy, I believe in basic HTML admin interfaces as those while "ugly" are far faster/more efficient than pretty frontend
But hey I'm not a web developer ((yet)) so if anyone knows of a way to make a speedy interface be my guest haha

**Comment by subject217** [comment]
tgui is for interfaces that benefit from having fancy schmancy outputs. progress bars that update without refreshing the page and stuff like that. its not useful for admin tools where the primary focus is usability and prettiness is secondary.

**Comment by Magikarpet** [comment]
you broke the vv dropdown

File: code/game/atoms_movable.dm [hunk 1]
```diff
@@ -735,7 +735,7 @@
 	// And animate the attack!
 	animate(I, alpha = 175, pixel_x = 0, pixel_y = 0, pixel_z = 0, time = 3)
 
-/atom/movable/vv_get_dropdown()
+/atom/movable/vv_get_dropdown_old()
```
Comment (optimumtact) [inline code/game/atoms_movable.dm]: this is sucky

File: code/modules/admin/callproc/callproc.dm [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+GLOBAL_PROTECT(AdminProcCaller)
+GLOBAL_VAR_INIT(AdminProcCallCount, 0)
+GLOBAL_PROTECT(AdminProcCallCount)
+GLOBAL_VAR(LastAdminCalledTargetRef)
+GLOBAL_PROTECT(LastAdminCalledTargetRef)
+GLOBAL_VAR(LastAdminCalledTarget)
+GLOBAL_PROTECT(LastAdminCalledTarget)
+GLOBAL_VAR(LastAdminCalledProc)
+GLOBAL_PROTECT(LastAdminCalledProc)
+GLOBAL_LIST_EMPTY(AdminProcCallSpamPrevention)
```
Comment (optimumtact) [inline code/modules/admin/callproc/callproc.dm]: can we stop with this shit, there is no point in both giving admins a foot gun and then trying to stop themselves shooting themselves in the foot.

It does not and never will make sense.

Admins can easily, easily break the game.

Comment (optimumtact) [reply code/modules/admin/callproc/callproc.dm]: this is literally all overhead for pointless reasons

File: code/modules/admin/view_variables/mass_edit_variables.dm [hunk 3]
```diff
@@ -1,265 +1,271 @@
...
-			if (typecache[thing.type])
-				. += thing
-			CHECK_TICK
-
-	else
-		for(var/datum/thing in world)
-			if (typecache[thing.type])
-				. += thing
-			CHECK_TICK
+/client/proc/cmd_mass_modify_object_variables(atom/A, var_name)
```
Comment (optimumtact) [inline code/modules/admin/view_variables/mass_edit_variables.dm]: did you really just respace this entire file?

File: code/modules/admin/view_variables/view_variables.dm [hunk 4]
```diff
@@ -0,0 +1,282 @@
+/client/proc/debug_variables(datum/D in world)
+	set category = "Debug"
+	set name = "View Variables"
+	//set src in world
+	var/static/cookieoffset = rand(1, 9999) //to force cookies to reset after the round.
+
+	if(!usr.client || !usr.client.holder) //The usr vs src abuse in this proc is intentional and must not be changed
```
Comment (optimumtact) [inline code/modules/admin/view_variables/view_variables.dm]: you should explain WHY, not what

File: code/modules/admin/view_variables/view_variables.dm [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+	var/hash
+
+	var/type = islist? /list : D.type
+
+	if(istype(D, /atom))
+		sprite = getFlatIcon(D)
+		hash = md5(sprite)
+		src << browse_rsc(sprite, "vv[hash].png")
+
+		/*		//getFlatIcon is expensive, this is the old code that just got the icon without using GFI
```
Comment (optimumtact) [inline code/modules/admin/view_variables/view_variables.dm]: dead code

File: code/modules/admin/view_variables/mass_edit_variables.dm (hunk 3 above)
Comment (kevinz000) [reply code/modules/admin/view_variables/mass_edit_variables.dm]: uhhh lemme check

**Comment by SpaceManiac** [comment]
looked at this for about 5 minutes
arbitrary `locate()` might be a security problem? needs to be checked
`continue` in do-while does check the condition so some of these do-whiles won't work right
size of this changeset scares me

**Comment by kevinz000** [comment]
@SpaceManiac uhh can i get the location of the locate() if it's in the vv do topic stuff in theory it'll already be checked for admin href soo

**Comment by kevinz000** [comment]
actually hm it's possible it can locate() world or un vv selectable datums but uhh wait one sec lemme go look

**Comment by AnturK** [comment]
So what's the state of this

File: code/__DEFINES/vv.dm [hunk 6]
```diff
@@ -19,3 +19,62 @@
...
+#define VV_HREF_TARGET_1V(target, href_key, text, varname) "<a href='[VV_HREF_TARGET_INTERNAL(target, href_key)];[VV_HK_VARNAME]=[varname]'>[text]</a>"		//for stuff like basic varedits, one variable
+
+#define GET_VV_TARGET locate(href_list[VV_HK_TARGET])
+#define GET_VV_VAR_TARGET href_list[VV_HK_VARNAME]
+
+//Helpers for vv_get_dropdown()
+#define VV_DROPDOWN_OPTION(href_key, name) . += "<option value='?_src_=vars;[HrefToken()];[href_key]=TRUE;target=[REF(src)]'>[name]</option>"
+
+//Helpers for vv_do_topic(list/href_list)
+#define IF_VV_OPTION(href_key) if(href_list[href_key])
```
Comment (AnturK) [inline code/__DEFINES/vv.dm]: Let's not.

**Review by AnturK — COMMENTED** [review]
I don't like that legacy dropdown, either keep it as it was or change everything.

**Pull request closed without merging**